src/xrpld/app/tx/detail/Credentials.cpp

```diff
@@ -0,0 +1,529 @@
+        msg.add32(HashPrefix::credential);
+        msg.addBitString(kCred.key);
+
+        if (!verify(pk, msg.slice(), *signature, /*canonical*/ true))
+        {
+            JLOG(j.trace()) << "Malformed transaction: bad signature.";
+            return temBAD_SIGNATURE;
+        }
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+CredentialCreate::preclaim(PreclaimContext const& ctx)
+{
+    auto const credType(ctx.tx[sfCredentialType]);
+    AccountID const account(ctx.tx[sfAccount]);
+    auto const subject = ctx.tx[~sfSubject].value_or(account);
+
+    std::optional<AccountID> signer = ctx.tx.isFieldPresent(sfIssuerPubKey)
+        ? std::optional<AccountID>(
+              calcAccountID(PublicKey(ctx.tx[sfIssuerPubKey])))
+        : std::nullopt;
+    auto const issuer = ctx.tx.isFieldPresent(sfIssuer)
+        ? ctx.tx[sfIssuer]
+        : (signer ? *signer : account);
+
+    if (ctx.tx.isFieldPresent(sfSubject) &&
+        !ctx.view.exists(keylet::account(subject)))
+    {
+        JLOG(ctx.j.trace()) << "Subject doesn't exist.";
+        return tecNO_TARGET;
+    }
+
+    if (ctx.view.exists(keylet::credential(subject, issuer, credType)))
+    {
+        JLOG(ctx.j.trace()) << "Credential doesn't exist.";
+        return tecDUPLICATE;
+    }
+
+    if (ctx.tx.isFieldPresent(sfIssuer))
+    {
+        // If Issuer present then IssuerPubKey must be the public key of the
+        // regular key for Issuer account. Let's check that it is true.
+
+        auto const sleIssuer = ctx.view.read(keylet::account(issuer));
+        if (!sleIssuer)
+        {
+            JLOG(ctx.j.trace()) << "Issuer doesn't exist.";
+            return tecNO_ISSUER;
+        }
+
+        if (sleIssuer->getAccountID(sfRegularKey) != signer)
+        {
+            JLOG(ctx.j.trace())
+                << "Regular key doesn't belong to Issuer account.";
+            return tecNO_REGULAR_KEY;
+        }
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialCreate::doApply()
+{
+    auto const sleOwner = view().peek(keylet::account(account_));
+
+    {
+        STAmount const reserve{view().fees().accountReserve(
+            sleOwner->getFieldU32(sfOwnerCount) + 1)};
+        if (mPriorBalance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    auto const subject = ctx_.tx[~sfSubject].value_or(account_);
+    auto const issuer = ctx_.tx.isFieldPresent(sfIssuer)
+        ? ctx_.tx[sfIssuer]
+        : (ctx_.tx.isFieldPresent(sfIssuerPubKey)
+               ? calcAccountID(PublicKey(ctx_.tx[sfIssuerPubKey]))
+               : account_);
+    auto const credType(ctx_.tx[sfCredentialType]);
+    Keylet const kCred = keylet::credential(subject, issuer, credType);
+    auto const sleCred = std::make_shared<SLE>(kCred);
+
+    auto const optExp = ctx_.tx[~sfExpiration];
+    if (optExp)
+    {
+        std::uint32_t const closeTime =
+            ctx_.view().info().parentCloseTime.time_since_epoch().count();
+
+        if (closeTime > *optExp)
```

Comment:
Every other time checks from other transactions  are in doApply, and I was told to put it here. tecEXPIRED  cause we don't return tem from doApply

Reply:
@mvadari @gregtatcam please collaborate together and decide where this check should be, because you give me controversial advises.

Reply:
I think @mvadari said the check should be done before the reserve check. Which is correct. But it doesn't mean that it should not be moved to `preclaim`.

Reply:
That is what @mvadari told me to do - move it from `preclaim` to `doApply`

Reply:
@mvadari why do you think it should be in `doApply`?

Reply:
I suggested it be in `doApply` because that seems to be the norm in this codebase (for example, with escrows). I assumed it was something related to timing of when it gets executed in consensus and maybe sometimes it's off. If there's no reason for that, then I'm fine moving it to `preclaim`.

Reply:
I see. Interestingly, `OfferCreate` does it in both `preclaim` and `doApply`. Perhaps you are right. It might the timing. And it's not expired in `preclaim` but expired in `doApply`. I'm sold on this.